### ISSUE TITLE
Doesn't prove the point

### ISSUE THREAD

**Issue by mvmn** [cmuratori/refterm#47]
The whole point of this project is to prove that terminal emulators can be fast if they want to.
But terminal emulators have to actually be useful. And for that purpose they have to support all the necessary features.
This program doesn't have much to offer in terms of the features. Thus it has the luxury of being fast that stems from the fact that it is very simple and feature-poor (well, feature-broke would be a better suiting terms, because it's really not doing much of use at all).

In order to prove the point this project has to be developed into a full-fledged terminal emulator on par or at least close feature-parity wise with other terminal emulators, and retain the speed of rendering while doing so. Only then it would prove anything.

**Comment by mmozeiko** [comment]
I think you missed the word "renderer" in About section of this project (on left side of github):

> Reference monospace terminal renderer

Point of this project was to show that rendering with high-performance arbitrary terminal text can be done with high-performance. And is not worthy of a PhD thesis. In no place this claims to do a fully functional terminal.

**Comment by mvmn** [comment]
> In no place this claims to do a fully functional terminal

I'm not saying it claims to be a full-fledged terminal emulator. I'm saying without actually doing that it proves no point.
A renderer can be fast on it's own, but to prove the point it has to work fast when integrated with the rest of the terminal emulator features.

**Comment by mmozeiko** [comment]
> A renderer can be fast on it's own

But that is exactly what was claimed to be not possible - rendering monospace text. That's why this was created.

Can you give example which feature of terminal emulator would be not possible to implement with current refterm's text rendering approach? Or why would it affect performance in any way?

**Comment by mvmn** [comment]
> But that is exactly what was claimed to be not possible

Claimed where by whom?

> Can you give example which feature

It's not about one specific feature, but about the **complexity** of a full-fledged solution.

**Comment by mmozeiko** [comment]
Claimed by Windows Terminal developers.

Nobody talked about complexity of full-fledged solution when "refterm" happened. Windows Terminal developers claimed that text rendering is slow and hard to do fast. refterm shows that this claim is false. That's all it is.

**Comment by mvmn** [comment]
> Claimed by Windows Terminal developers

Where is that claim? Can I see a quote?

**Comment by mmozeiko** [comment]
The original issue where conversation started: `https://github.com/microsoft/terminal/issues/10362#issuecomment-862844333`

**Comment by mvmn** [comment]
Well, they're asking some questions there, but not making much of a statement - beside complaining on the tone of comments.

**Comment by christianparpart** [comment]
I would like to raise the actual point of @mvmn here. Creating fast renders is trivial for as long as it doesn't do anything but simply render monospaced text. That point was also proven by this little project here. However, @mvmn has an actual point here. 

That is, you cannot simply create a fast monospaced renderer without completely ignoring the backend it is supposed to be connected to. I was once asking cmuratori in one of his Twitch streams how he solves the challenge of

- multiple screen buffers, as well as
- columnar and rectangular mutation VT sequences, or 
- things as trivial as auto-wrapping text vs non-wrapped text (see SM VT sequence for the modes).
- more complex VT sequences, such as inlined images (Sixel), file transfer, SM/DECSM, SCS, ...
- handle proper grapheme cluster segmentation in the context of proper terminal grid positioning

The answer was, he doesn't know.

If you look closely at refterm's source code, you even see that it's cutting lines hard-coded at 4k bytes, so the text to be printed is not equivalent to what the user expected to print. It ends up being forged output.

The other problem with the backend buffer implementation is, that it would not be a wise choice as it might contain payload data that is not even meant to be printed on the screen, such as clipboard pasts, or file transfer data (VT extension), or Sixel payload data for rendering images on the screen. 

The basic idea behind this repository here is to proof fast rendering. And for that, it chose a backend model where it reduces the amount of copies as much as possible, which is a great aspect and I tried really hard about thinking how **I** myself would be able to incorporate such an idea into an actual terminal emulator (that I am authoring and maintaining since some years already). Turns out it is impossible in the way refterm does, due to the incomplete list of the problems as mentioned above.

The actual reason why most terminals are slow is not even the renderer, in fact, but the way how VT sequences are interpreted. refterm itself does not process any VT sequence at all unless it has to render something on the screen. That is far too late in the process, because there might have been state changes. This is a lateral design failure, while it would actually be great if it would be possible the way he did it. But truth is, this design is simply not working. (hance, refterm's outcome cannot be compared with actual TEs)

In the end, no real terminal will come even close to what refterm wanted to prove. To make it a fair point, objectively speaking, the base requirements must be set **AND EQUALLY** implemented on all sides.

p.s.: an actual terminal emulator should not try to render as many frames per second as possible but as little as required, to save energy resources. This is not a game, but a systems software. In other words, only thinking about the renderer is the wrong place to optimize for. :)

**Comment by cmuratori** [comment]
I have had to deal with a lot of ignorance because of this project, but this is probably one of the most ignorant comments I have received.

1. You seem to think refterm was created to "prove a point" about parsing, when in fact, the parser is only there for demonstration of the renderer. The original thread which refterm was created in response to _was a thread claiming Windows Terminal was slow because of its renderer_. Nobody was talking about VT codes beyond just the fact that the _renderer_ had to handle different foreground and background colors.
2. I have already answered the question of "potentially slow-to-parse VT codes" [here](https://github.com/cmuratori/refterm/issues/48). They should not - and in a properly written terminal, would not - ever slow down the common cases people use, all of which _are_ actually supported by refterm at high bandwidth (colors, etc.). Why people think adding rarely-used features requires slowing down your fast path is beyond me. If you can't figure out how to make a particular rarely-used feature go fast, you just put that on the slow path, and downshift when you see it. That said, I don't know that there even _is_ a slow path here - it's highly likely that it is simply people assuming there is. But since refterm was a rendering example and not a VT parsing example, I was not asking that particular question, and so did not spend the time to work it out.
3. Refterm uses less resources than Windows terminal, and only produces frames as fast as is necessary to display new information to the user. When nothing is changing, it does not spend any CPU time. So I have no idea what your "PS" is even referring to. As per the original thread that spawned this project, the _entire point_ was that people were claiming high frame rates were impossible with a terminal _because of monospace font rendering_. This project was designed to conclusively demonstrate that monospace font rendering is not a bottleneck.

TL;DR: The idea that refterm's speeds are unachievable on regular console output because of esoteric VT codes is absurd. If someone thinks that's true, their opinion on programming should not be taken seriously. Refterm already parses VT codes and, at no speed cost, could trivially downshift to a different parser - even a different _renderer_, if so desired - the instant it saw one it didn't handle in the fast path parser. That people should have to suffer with slow terminal throughput simply because programs _they never use_ might output a rare, slow VT code is just another example of how people don't understand even basic programming concepts anymore, like fast path/slow path.

In closing, if this had been a VT-parsing project, I would have gone and analyzed all the VT codes and made something that parses and processes them as quickly as possible, compliant to spec. I wouldn't have done a renderer. Then, deeply ignorant people like this would have shown up in the comments to _that_ project, and explained how it didn't prove anything about anything, because everyone knows that rendering fonts is the really hard problem, and that terminals "aren't slow because of VT code parsing".

The state of modern programming discourse is an utter disaster. Terminals are not hard problems. Stop pretending they are.

\- Casey
